src/modules/aws_ec2/mappers/instance.ts

```diff
@@ -290,6 +331,81 @@ export class InstanceMapper extends MapperBase<Instance> {
     _res => undefined,
   );
 
+  // given an instance reads the mapping from the associate AMI , match
+  // it with the current mapped volumes and generate the final mapping
+  async generateBlockDeviceMapping(
+    ctx: Context,
+    ami: string,
+    maps: InstanceBlockDeviceMapping[],
+    instance: Instance,
+    encrypted: boolean,
+  ) {
+    // start reading the block device mapping from the image
+    const client = await ctx.getAwsClient(instance.region);
+    const amiImage = (await this.describeImages(client.ec2client, [ami]))?.Images?.pop();
+
+    if (amiImage) {
+      const imageMapping = amiImage.BlockDeviceMappings;
+
+      // check if there is any mapped volume that doesn't exist on instance mapping, and error
+      // or if there is any volume mapped that is not set as root
+      for (const dev of maps ?? []) {
+        // try to find the device name on instance mapping
+        const vol = imageMapping?.find(item => item.DeviceName === dev.deviceName);
+        if (!vol) throw new Error('Error mapping volume to a device that does not exist for the AMI');
+        if (!dev.volume?.isRootDevice) throw new Error('Error mapping volume that is not root');
+      }
+      const region = instance.region;
+      for (const map of imageMapping ?? []) {
+        // check if there is an associated volume for that instance, volume and device name
+        const vol = maps?.find(item => item.deviceName === map.DeviceName);
+        if (vol) {
+          if (vol.volumeId) {
+            if (vol.volume) {
+              // map it to the ebs mapping
+              if (!vol.volume.isRootDevice) throw new Error('Error mapping volume that is not root');
+
+              let snapshotId;
+              if (vol.volume.snapshotId && instance.region === vol.region) snapshotId = vol.volume.snapshotId;
+              else snapshotId = map.Ebs?.SnapshotId;
```

Comment:
can we refactor this using early returns to avoid the nested conditionals to make it easier to follow?